### PR TITLE
completion: Refactor proxy completion logic in a new package

### PR DESCRIPTION
**Pull request by rlenglet**
Move the completion into its own package: github.com/cilium/cilium/pkg/completion.
Rename CompletionContainer into WaitGroup to reflect the similarity with sync.WaitGroup.
Refactor Completion and WaitGroup to take a Context and handle context cancellation.
Rename Completion.Completed into Complete to make it a verb.
Added Completion.Completed method to return a channel, to make it easier to use in unit tests.

Revert the (de)serialization of the ProxyCompletions field to/from JSON within the Endpoint struct, and rename ProxyCompletions into ProxyWaitGroup.

Signed-off-by: Romain Lenglet <romain@covalent.io>

This is a follow up to https://github.com/cilium/cilium/pull/2694.

### CODE REVIEW

**Comment by rlenglet** [comment]
test-me-please

**Comment by ianvernon** [comment]
@rlenglet would you be alright with adding yourself as a CODEOWNER of this new package in CODEOWNERS?

**Comment by rlenglet** [comment]
@ianvernon sure, I added this: `+pkg/completion/ @cilium/proxy` since this is currently mostly used for proxy interactions.

**Comment by rlenglet** [comment]
test-me-please

**Comment by rlenglet** [comment]
test-me-please

**Comment by rlenglet** [comment]
test-me-please

File: pkg/completion/completion.go
Comment (nebril) [inline pkg/completion/completion.go]: I think using reflection in WaitGroups is not needed at all, `Wait` logic can be done by multiplexing completion channels into one channel and selecting from it and `wg.ctx.Done()`. Example implementation is here: https://play.golang.org/p/fHnkmHm4q8s

Please let me know if this makes sense. I wrote above snippet because I think that reflection should be used as little as possible, and I think that in this case implementation with concurrency primitives is the same in terms of complexity (and data stored in waitgroups is less complex, as we only need a channel per completion).

If you feel that this is adding unnecessary complexity to the code please dismiss review.

**Review by nebril — DISMISSED** [review]
General comment against using reflection with example of how it could be done without it inline.

**Comment by rlenglet** [comment]
test-me-please

File: pkg/completion/completion.go
Comment (rlenglet) [reply pkg/completion/completion.go]: @nebril Thanks for the review! I also would very much like to get rid of `reflect.Select`, especially since it's quite expensive.
However, any solution with multiple writers into a single channel makes it really hard to safely determine when to close that channel. The general rule is that a channel must be closed by its writer, since only the writer know when it stopped writing and writing into a closed channel panics. It's becoming hard when you have multiple identical writers.
I noticed that you don't close the multiplex channel in your solution.

Closing in `Wait` may work. It may look like this:
```
wg.counterLocker.Lock()
defer wg.counterLocker.Unlock()
defer close(multiplex)
for count > 0 {
    select {
    case <-multiplex:
        count--
    case <-ctx.Done():
        for _, comp := range wg.pendingCompletions {
            comp.Completed()
	}
        for count > 0 {
            <-multiplex
            count--
        }
	return wg.ctx.Err()
    }
    return nil
}
```

I'll try that.
Did I miss something?

Comment (rlenglet) [reply pkg/completion/completion.go]: With that implementation, I could even get rid of the individual channel per completion, and have them write directly into the "multiplex" channel, since the implementation guarantees that we close it only after all completions are completed. So a single channel is sufficient.

Comment (rlenglet) [reply pkg/completion/completion.go]: Hmm I may need to keep individual goroutines, because the multiplex channel will remain of a bounded size, so the loop to call `comp.Completed()` may block indefinitely otherwise.
I could simply have the implementation of `Completed()` start a goroutine to write into the multiplex channel.

Comment (rlenglet) [reply pkg/completion/completion.go]: Or I could simply start a goroutine for the loop itself:
```
wg.counterLocker.Lock()
defer wg.counterLocker.Unlock()
defer close(multiplex)
for count > 0 {
    select {
    case <-multiplex:
        count--
    case <-ctx.Done():
        go func() {
            for _, comp := range wg.pendingCompletions {
                comp.Completed()
	    }
        }()
        for count > 0 {
            <-multiplex
            count--
        }
	return wg.ctx.Err()
    }
    return nil
}
```

Comment (rlenglet) [reply pkg/completion/completion.go]: I found a much better and simpler implementation of `Wait` without any additional channel or goroutine:
```
wg.counterLocker.Lock()
defer wg.counterLocker.Unlock()
for _, comp := range wg.pendingCompletions {
    select {
    case <-comp.done:
        continue
    case <-wg.ctx.Done():
        for _, comp := range wg.pendingCompletions {
            comp.Completed()
	}
	return wg.ctx.Err()
    }
}
return nil
```
I can now get rid of the select cases.

Comment (nebril) [reply pkg/completion/completion.go]: Nice!

Comment (rlenglet) [reply pkg/completion/completion.go]: Done. @nebril please review the latest update.

**Comment by rlenglet** [comment]
test-me-please

**Comment by ianvernon** [comment]
Can you add a `doc.go` file? See `pkg/node/doc.go` as an example.

File: pkg/endpoint/endpoint.go
Comment (ianvernon) [inline pkg/endpoint/endpoint.go:341]: Is this also protected by the BuildMutex? If so, I'd like to keep that comment.

Comment (rlenglet) [reply pkg/endpoint/endpoint.go:341]: It doesn't need to.

Comment (ianvernon) [reply pkg/endpoint/endpoint.go:341]: OK, thanks.

**Comment by rlenglet** [comment]
@ianvernon, I added doc.go.

**Comment by rlenglet** [comment]
test-me-please

**Comment by rlenglet** [comment]
test-me-please

**Review by nebril — APPROVED** [review]

**Review by vadorovsky — APPROVED** [review]

**Review by raybejjani — CHANGES_REQUESTED** [review]
Would it be ok if we flattened the interfaces into the types that implement them? Unless you had a larger scheme I'm not clear on why the interfaces are needed (I didn't check for an import loop but I'm really hoping that's not the case).

More constructively, is the intention just to mimic `sync.WaitGroup` with a ctx based timeout? or allow a "Done" call on each completion?
If it's the former (mimic `WaitGroup`), why not just wrap `WaitGroup` itself with a ctx?
If it's the latter (independent completions), what was the impetus here? I can see the argument against leaking a waitgroup when a ctx timeout occurs but this implementation also "leaks" in the same way.

**Comment by rlenglet** [comment]
test-me-please

**Comment by rlenglet** [comment]
@raybejjani 
> Would it be ok if we flattened the interfaces into the types that implement them? Unless you had a larger scheme I'm not clear on why the interfaces are needed (I didn't check for an import loop but I'm really hoping that's not the case).

I need to be able to mock those, especially `Completion`, because otherwise it makes unit tests significantly more complicated. I have coming changes where I mock `Completion`.

> More constructively, is the intention just to mimic sync.WaitGroup with a ctx based timeout? or allow a "Done" call on each completion?

The motivation is the combination of:
1. Having a completion callback per concurrent computation.
2. Being able to mock the completion callback. `sync.WaitGroup` is not an interface, so it can't be mocked. `Completion` can.
3. Select a `Context.Done()` in `Wait`, and make the `Context` available to concurrent calls, so that they can cancel early if the `Context` is canceled.

I thought that the interface was similar enough to `sync.WaitGroup` that it would make sense to use the same name.
If you find the "WaitGroup" name confusing, I can change it.

> If it's the former (mimic WaitGroup), why not just wrap WaitGroup itself with a ctx?
> If it's the latter (independent completions), what was the impetus here? I can see the argument against leaking a waitgroup when a ctx timeout occurs but this implementation also "leaks" in the same way.

I don't understand this statement. What I am leaking in this implementation?

**Comment by rlenglet** [comment]
@raybejjani The problem with canceling a call to `sync.WaitGroup.Wait` when a `Context` is canceled, is that it makes the design much more complicated:
* The call to `sync.WaitGroup.Wait` has to be done in a separate goroutine.
* In order to unblock that call, one has to make a number of calls to `sync.WaitGroup.Done`. The problem is determining the exact number of calls to make. It has to be the exact number of pending completions. If we don't do enough calls, the call remains blocked, if we do even one extra call, `WaitGroup` panics.
* Tracking the exact number of pending completions would require implementing pretty much what I implement in this PR.

Overall, an implementation wrapping a `sync.WaitGroup` would be much more complicated, it would use an extra goroutine, and would bring no benefit compared to my solution.

File: pkg/completion/doc.go
Comment (ianvernon) [inline pkg/completion/doc.go:57]: This is great documentation.

**Review by ianvernon — APPROVED** [review]

**Comment by rlenglet** [comment]
test-me-please

**Comment by rlenglet** [comment]
Rebased.

**Comment by rlenglet** [comment]
test-me-please

**Comment by raybejjani** [comment]
These responses are a little out of order to avoid confusion.

>>If it's the latter (independent completions), what was the impetus here? I can see the argument against leaking a waitgroup when a ctx timeout occurs but this implementation also "leaks" in the same way.

>I don't understand this statement. What I am leaking in this implementation?

I was wrong. I didn’t notice the handling you had (on pkg/completion/completion.go:96) where you cleared the remaining completions. In fact, that loop is quite excellent :)
The leak I had in mind is the one you’re accounting for: if I cancel a context I will have a goroutine stuck in sync.WaitGroup.Wait() (which is there just to allow us to do a select on ctx.Done() )

>I need to be able to mock those, especially Completion, because otherwise it makes unit tests significantly more complicated. I have coming changes where I mock Completion.

I’m guessing the mock is in https://github.com/cilium/cilium/pull/2764 but I don’t really understand why we need to mock it. The WaitGroup is such a straightforward interface that we could use it directly, no? In fact, because we have the ctx available we can ensure everything is cancelled on test exit.
Nonetheless, I see that you have a checker built around the mock in that test. I would argue you could do the same by adding a iscompleted function to the WaitGroup directly but that’s not a reason on it’s own to change what you have.


>The motivation is the combination of:
>1. Having a completion callback per concurrent computation.
>2. Being able to mock the completion callback. `sync.WaitGroup` is not an interface, so it can't be >mocked. `Completion` can.
>3. Select a `Context.Done()` in `Wait`, and make the `Context` available to concurrent calls, so that they can cancel early if the `Context` is canceled.
>
>I thought that the interface was similar enough to `sync.WaitGroup` that it would make sense to use the same name.
>If you find the "WaitGroup" name confusing, I can change it.

On the contrary! I think WaitGroup is the best name since it’s intended to be used that way. I think the main divergence is that I didn’t expect to mock this type the way you have. I also envisioned that a much more naive wrapping of sync.WaitGroup was possible.

Anyway, I don't want to block this PR since my objections are more style and philosophy than an actual broken thing.

**Review by raybejjani — APPROVED** [review]

File: pkg/completion/completion.go
Comment (raybejjani) [inline pkg/completion/completion.go]: Oh, if you wanted to avoid the lock here, and the completed variable, we could do this:
```golang
	select {
	case <-c.done:
	default:
		close(c.done)
	}
```

Comment (rlenglet) [reply pkg/completion/completion.go]: Beautiful! Done.

**Comment by rlenglet** [comment]
@raybejjani I removed the interfaces, and directly named the structs `WaitGroup` and `Completion`. It's looking quite good. I added a `Completed()` method into `Completion` to return the channel. That's all I need to use in unit test, so that's good for me.

**Comment by rlenglet** [comment]
test-me-please

**Comment by rlenglet** [comment]
test-me-please

**Comment by rlenglet** [comment]
test-me-please

File: pkg/endpoint/endpoint.go
Comment (jrajahalme) [inline pkg/endpoint/endpoint.go]: Not your code, but while debugging something else yesterday I noticed that we should use `e.getLogger()' here.

File: pkg/envoy/envoy_test.go
Comment (jrajahalme) [inline pkg/envoy/envoy_test.go]: Add "after stopping Envoy" to the end.

File: pkg/endpoint/endpoint.go
Comment (jrajahalme) [reply pkg/endpoint/endpoint.go:341]: IMO ProxyWaitGroup has to be protected by BuildMutex as that is what all the call sites are doing, and they would break if we add new code that overwrites ProxyWaitGroup while it is being used by the existing call sites.

Comment (jrajahalme) [reply pkg/endpoint/endpoint.go:341]: This may not apply for a unit test case, but that's beside the point.

File: pkg/completion/completion.go
Comment (jrajahalme) [inline pkg/completion/completion.go]: In order to enforce the semantics in the comment it might be clearer to make to take the WaitGroup as a receiver so that the context would not need to be explicitly passed, i.e.:
```
func (wg *WaitGroup) NewCompletion() *Completion {
...
}
```

Comment (jrajahalme) [reply pkg/completion/completion.go]: This does not need to be exported.

File: pkg/completion/doc.go
Comment (jrajahalme) [inline pkg/completion/doc.go:31]: Needs an update as it is not an interface any more.

File: pkg/completion/completion.go
Comment (jrajahalme) [inline pkg/completion/completion.go]: Completion.Complete() being idempotent each call to AddCompletion() must return a pointer to an unique/new Completion object. It does, but it needs to be part of the interface/API/docs now.

**Review by jrajahalme — CHANGES_REQUESTED** [review]
Great work! More robust than my first cut :-)

There is one comment about the BuildMutex and a needed doc update that I feel must be addressed, rest is up to your discretion.

File: pkg/endpoint/endpoint.go
Comment (rlenglet) [reply pkg/endpoint/endpoint.go]: Done.

File: pkg/envoy/envoy_test.go
Comment (rlenglet) [reply pkg/envoy/envoy_test.go]: Done.

File: pkg/completion/completion.go
Comment (rlenglet) [reply pkg/completion/completion.go]: The whole point of this function was to be able to use a Completion in unit tests without creating a WaitGroup. What you propose is nearly identical to AddCompletion. I'll remove it since it seems to be confusing.

File: pkg/completion/doc.go
Comment (rlenglet) [reply pkg/completion/doc.go:31]: Done.

File: pkg/completion/completion.go
Comment (rlenglet) [reply pkg/completion/completion.go]: OK, clarified that.

File: pkg/endpoint/endpoint.go
Comment (rlenglet) [reply pkg/endpoint/endpoint.go:341]: I haven't removed any locking using BuildMutex. Whatever locking was done is still there.
Sorry I might have misunderstood @ianvernon's original question.

**Comment by rlenglet** [comment]
test-me-please

File: pkg/endpoint/endpoint.go
Comment (jrajahalme) [reply pkg/endpoint/endpoint.go:341]: This comment thread is due to your removal of wording "protected by the BuildMutex!" from the comment. It needs to be brought back to help avoid races in future.

**Comment by jrajahalme** [comment]
The comment on `ProxyWaitGroup` in `Endpoint` needs to retain the language that it is protected by BuildMutex. This PR removes that, and it needs to be brought back to help avoid races in future.

**Comment by rlenglet** [comment]
test-me-please

**Comment by rlenglet** [comment]
Re-added a comment about BuildMutex.

**Review by jrajahalme — APPROVED** [review]

**Pull request merged**
